### PR TITLE
feature_ip_click_counter

### PR DESCRIPTION
**Pull request by tspurway**
This is a 'fraud detector'.  It works by running an Inferno job daily (1am) on the previous day's click/impression stream data and counting excessively repetitive clicks from specific IP addresses.  

It uses datadog statsd events to report.

r? @oyiptong

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -9,7 +9,7 @@
     author='tspurway',
     author_email='tspurway@mozilla.com',
     description='Inferno rules for Tiles Project',
-    install_requires=['psycopg2', 'inferno', 'boto', 'geoip2', 'ua_parser', 'statsd', 'disco'],
+    install_requires=['psycopg2', 'inferno', 'boto', 'geoip2', 'ua_parser', 'statsd', 'disco', 'datadog'],
```
Comment (oyiptong) [inline setup.py]: the setup script attempts to install disco form pypi which resolves to this project: https://pypi.python.org/pypi/Disco

That's probably not what you mean to install

Comment (tspurway) [inline setup.py]: Good catch.  It's interesting that the disco client isn't in PyPI.  fixing....

**Comment by tspurway** [comment]
@oyiptong I am switching the r? 

r? @ncloudioj

**Comment by ncloudioj** [comment]
Sure, I am on it.

File: infernyx/rules.py [hunk 2]
```diff
@@ -231,6 +232,44 @@ def report_rule_stats(job):
...
+    ips = []
+    for (_, ip), (total,) in it:
+        if total_ips >= 1000:
+            break
+        ips.append("%20s %d" % (ip, total))
+        total_ips += 1
+
+    if total_ips:
+        msg = '\n'.join(ips)
+        statsd.event("Suspicious IP report", msg)
```
Comment (ncloudioj) [inline infernyx/rules.py]: The `statsd.event()` will raise an exception if the length of msg is greater than 8k. We'd better drop an exception handler here.

Comment (tspurway) [inline infernyx/rules.py]: can you doublecheck the 8k limit for datadog?  that kind of sucks.

Comment (ncloudioj) [inline infernyx/rules.py]: yup. I've checked its source code. That limit is weird though. Can we send
multiple msgs instead?

Tim Spurway notifications@github.com于2015年5月13日星期三写道：

**Comment by tspurway** [comment]
This feature has been refactored.  Instead of putting suspicious IPs in a statsd message, we are writing them to a table, along with the current date.  A daily job will run to 'negate' blacklisted IPs for the previous day in impression_stats_daily.  Also, any IP that hasn't been 'seen' in the past 7 days will be deleted.

Note this feature is dependent on migrations in https://github.com/oyiptong/splice/pull/63
r?   @ncloudioj

File: infernyx/rules.py [hunk 3]
```diff
@@ -297,9 +343,49 @@ def tag_results(suffix, job):
...
+        time_delta={'oclock': 4},
+        rule_cleanup=report_rule_stats,
+        map_input_stream=chunk_json_stream,
+        map_init_function=impression_stats_init,
+        blacklisted=True,
+        parts_preprocess=[clean_data, parse_date, parse_locale, parse_ip, filter_blacklist, parse_ua, parse_tiles],
+        geoip_file=GEOIP,
+        partitions=32,
+        sort_buffer_size='25%',
+        locale_whitelist={'ach', 'af', 'an', 'ar', 'as', 'ast', 'az', 'be', 'bg', 'bn-bd', 'bn-in', 'br', 'bs',
```
Comment (ncloudioj) [inline infernyx/rules.py]: There is another copy of `locale_whitelist` in rule `impression_stats`, better take it out and reuse it?

Comment (tspurway) [inline infernyx/rules.py]: Good catch, Nan.  I fixed that in the latest push

File: infernyx/rules.py [hunk 4]
```diff
@@ -359,11 +445,33 @@ def tag_results(suffix, job):
...
+                                 database=RS_DB,
+                                 user=RS_USER,
+                                 password=RS_PASSWORD,
+                                 bucket_name=RS_BUCKET),
+        partitions=32,
+        sort_buffer_size='25%',
+        combiner_function=combiner,
+        key_parts=['date', 'ip'],
+        value_parts=['impressions', 'clicks'],
+        column_mappings={'impressions': None, 'clicks': None},
```
Comment (ncloudioj) [inline infernyx/rules.py:468]: This line appears to be redundant, inferno will ignore the column mapping if its value is `None`

**Comment by ncloudioj** [comment]
r+

File: infernyx/rules.py (hunk 4 above)
Comment (tspurway) [inline infernyx/rules.py:468]: Inferno used to ignore columns when their column mappings were set to
None.  That isnt happening anymore!

I'll submit a fix.

**Pull request closed without merging**